### PR TITLE
Refactorings

### PR DESCRIPTION
**Pull request by traiansf**

### CODE REVIEW

**Comment by raduom** [comment]
There are no comments in the code.

File: src/haskell/parser/src/KoreAST.hs [hunk 1]
```diff
@@ -3,119 +3,163 @@ module KoreAST where
...
     deriving (Show, Eq)
 
 data Pattern
     = AndPattern
-        { andPatternSort   :: !Sort
+        { andPatternSort   :: !ObjectSort
         , andPatternFirst  :: !Pattern
         , andPatternSecond :: !Pattern
         }
     | ApplicationPattern
```
Comment (raduom) [inline src/haskell/parser/src/KoreAST.hs:97]: Combining sums with records is somewhat frowned upon (see https://stackoverflow.com/questions/37652243/record-syntax-and-sum-types for details).

File: src/haskell/parser/src/KoreLexemeImpl.hs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+skipWhitespace = whitespaceChunk
+
+lexeme :: Parser a -> Parser a
+lexeme p = p <* skipWhitespace
+
+genericIdParser :: CharSet -> CharSet -> (String -> a) -> Parser a
+genericIdParser firstCharSet charSet constructor = do
+    c <- Parser.peekChar'
+    idChar <- if not (c `CharSet.elem` firstCharSet)
+        then fail ("genericidParser: Invalid first character '" ++ c : "'.")
```
Comment (raduom) [inline src/haskell/parser/src/KoreLexemeImpl.hs:81]: Maybe this is a bit off, but have you considered providing context information when parsing fails? It does not look like there is any support for that in this parser.

Comment (virgil-serbanuta) [reply src/haskell/parser/src/KoreLexemeImpl.hs:81]: We considered it, but we assumed that this parser will work on generated code, so we decided it was low priority. I hope we will not regret it.

**Review by virgil-serbanuta — APPROVED** [review]

**Pull request merged**